tests/testmain.rs

```diff
@@ -1,78 +1,22 @@
 extern crate lal;
 
-extern crate fs_extra;
 #[macro_use] extern crate log;
-extern crate loggerv;
 extern crate parameterized_macro;
 extern crate tempdir;
-extern crate walkdir;
-
-use std::{
-    env, fs,
-    fs::File,
-    io::prelude::*,
-    path::{Path, PathBuf},
-    process::Command,
-    sync::Once,
-};
 
-use fs_extra::dir::{copy, CopyOptions};
+use std::{fs, path::Path};
+
 use parameterized_macro::parameterized;
-use tempdir::TempDir;
-use walkdir::WalkDir;
 
 use lal::*;
-use loggerv::init_with_verbosity;
-
-struct TestState {
-    backend: LocalBackend,
-    testdir: PathBuf,
-
-    // Keep the tempdir with TestState.
-    // The directory will be cleaned with the TestState is `Drop`ed.
-    tempdir: TempDir,
-}
-
-static START: Once = Once::new();
-
-fn setup() -> TestState {
-    START.call_once(|| {
-        env::set_var("SSL_CERT_FILE", "/etc/ssl/certs/ca-certificates.crt");
-        env::set_var("LAL_CODE_DIR", env::current_dir().unwrap().as_os_str());
-
-        // print debug output and greater from lal during tests
-        init_with_verbosity(2).expect("Setting up test logging");
-    });
-
-    let mut demo_config = PathBuf::from(env::var("LAL_CODE_DIR").unwrap());
-    demo_config.push("./configs/demo.json");
-
-    let mut testdir = PathBuf::from(env::var("LAL_CODE_DIR").unwrap());
-    testdir.push("tests");
-
-    // Do all lal tests in a tempdir as it messes with the manifest
-    let tempdir = TempDir::new("laltest").unwrap();
-
-    let backend = configure_local_backend(&demo_config, &tempdir.path());
-
-    TestState {
-        backend,
-        tempdir,
-        testdir,
-    }
-}
 
-// Copies the component to a temporary location for this test
-// and sets the working directory to that location
-fn clone_component_dir(component: &str, state: &TestState) -> PathBuf {
-    let copy_options = CopyOptions::new();
-
-    let from = state.testdir.join(component);
-    let to = state.tempdir.path().join(component);
-
-    copy(&from, state.tempdir.path(), &copy_options).expect("copy component to tempdir");
-    return to;
-}
+mod common;
```

Comment:
don't you feel this feels a bit awkward to have the logic for `setup` inside the module above then run them all the same in this outer file? a lot of this here strikes me a bit as a custom test harness

Reply:
I tried `rstest`, which has fixtures but that was incompatible with `serial`. I might try it again.

`speculate` has a `before()` syntax similar to rspec.
I haven't tried that yet, but it looks like it could work too.

Know of any others?

Reply:
I've now moved all of the checks to their own (command specific) files. `tests/testmain.rs` now only has four lines in it, and isn't a wrapper for other tests anymore.